<!DOCTYPE html>
<html lang="en">
<head>
    
    <meta charset="utf-8">
    <title>readers/mtx.js - Documentation</title>
    
    
    <script src="scripts/prettify/prettify.js"></script>
    <script src="scripts/prettify/lang-css.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc.css">
    <script src="scripts/nav.js" defer></script>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>

<input type="checkbox" id="nav-trigger" class="nav-trigger" />
<label for="nav-trigger" class="navicon-button x">
  <div class="navicon"></div>
</label>

<label for="nav-trigger" class="overlay"></label>

<nav >
    
    <input type="text" id="nav-search" placeholder="Search" />
    
    <h2><a href="index.html">Home</a></h2><h2><a href="https://github.com/LTLA/bakana" target="_blank" class="menu-item" id="repository" >Github repo</a></h2><h3>Classes</h3><ul><li><a href="AdtNormalizationState.html">AdtNormalizationState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="AdtNormalizationState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="AdtNormalizationState.html#fetchExpression">fetchExpression</a></li><li data-type='method' style='display: none;'><a href="AdtNormalizationState.html#summary">summary</a></li></ul></li><li><a href="AdtPcaState.html">AdtPcaState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="AdtPcaState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="AdtPcaState.html#summary">summary</a></li></ul></li><li><a href="AdtQualityControlState.html">AdtQualityControlState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="AdtQualityControlState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="AdtQualityControlState.html#summary">summary</a></li></ul></li><li><a href="BatchCorrectionState.html">BatchCorrectionState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="BatchCorrectionState.html#compute">compute</a></li></ul></li><li><a href="CellFilteringState.html">CellFilteringState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="CellFilteringState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="CellFilteringState.html#fetchFilteredAnnotations">fetchFilteredAnnotations</a></li><li data-type='method' style='display: none;'><a href="CellFilteringState.html#fetchFilteredBlock">fetchFilteredBlock</a></li><li data-type='method' style='display: none;'><a href="CellFilteringState.html#summary">summary</a></li><li data-type='method' style='display: none;'><a href="CellFilteringState.html#undoFiltering">undoFiltering</a></li></ul></li><li><a href="CellLabellingState.html">CellLabellingState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="CellLabellingState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="CellLabellingState.html#summary">summary</a></li></ul></li><li><a href="ChooseClusteringState.html">ChooseClusteringState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="ChooseClusteringState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="ChooseClusteringState.html#fetchClusterIndices">fetchClusterIndices</a></li><li data-type='method' style='display: none;'><a href="ChooseClusteringState.html#summary">summary</a></li></ul></li><li><a href="CombineEmbeddingsState.html">CombineEmbeddingsState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="CombineEmbeddingsState.html#compute">compute</a></li></ul></li><li><a href="CustomSelectionsState.html">CustomSelectionsState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#.computeVersusCustom">computeVersusCustom</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#addSelection">addSelection</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#computeVersus">computeVersus</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#fetchResults">fetchResults</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#fetchSelectionIndices">fetchSelectionIndices</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#fetchSelections">fetchSelections</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#removeSelection">removeSelection</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#summary">summary</a></li></ul></li><li><a href="FeatureSelectionState.html">FeatureSelectionState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="FeatureSelectionState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="FeatureSelectionState.html#summary">summary</a></li></ul></li><li><a href="H5adDataset.html">H5adDataset</a><ul class='methods'><li data-type='method' style='display: none;'><a href="H5adDataset.html#.format">format</a></li><li data-type='method' style='display: none;'><a href="H5adDataset.html#.unserialize">unserialize</a></li><li data-type='method' style='display: none;'><a href="H5adDataset.html#abbreviate">abbreviate</a></li><li data-type='method' style='display: none;'><a href="H5adDataset.html#annotations">annotations</a></li><li data-type='method' style='display: none;'><a href="H5adDataset.html#clear">clear</a></li><li data-type='method' style='display: none;'><a href="H5adDataset.html#load">load</a></li><li data-type='method' style='display: none;'><a href="H5adDataset.html#serialize">serialize</a></li></ul></li><li><a href="InputsState.html">InputsState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="InputsState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="InputsState.html#enableDatasetCache">enableDatasetCache</a></li><li data-type='method' style='display: none;'><a href="InputsState.html#fetchAnnotations">fetchAnnotations</a></li><li data-type='method' style='display: none;'><a href="InputsState.html#setDirectSubset">setDirectSubset</a></li><li data-type='method' style='display: none;'><a href="InputsState.html#summary">summary</a></li><li data-type='method' style='display: none;'><a href="InputsState.html#undoSubset">undoSubset</a></li></ul></li><li><a href="KmeansClusterState.html">KmeansClusterState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="KmeansClusterState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="KmeansClusterState.html#summary">summary</a></li></ul></li><li><a href="MarkerDetectionState.html">MarkerDetectionState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="MarkerDetectionState.html#.computeVersusCustom">computeVersusCustom</a></li><li data-type='method' style='display: none;'><a href="MarkerDetectionState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="MarkerDetectionState.html#computeVersus">computeVersus</a></li><li data-type='method' style='display: none;'><a href="MarkerDetectionState.html#fetchGroupResults">fetchGroupResults</a></li><li data-type='method' style='display: none;'><a href="MarkerDetectionState.html#numberOfGroups">numberOfGroups</a></li><li data-type='method' style='display: none;'><a href="MarkerDetectionState.html#summary">summary</a></li></ul></li><li><a href="NeighborIndexState.html">NeighborIndexState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="NeighborIndexState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="NeighborIndexState.html#summary">summary</a></li></ul></li><li><a href="NormalizationState.html">NormalizationState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="NormalizationState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="NormalizationState.html#fetchExpression">fetchExpression</a></li><li data-type='method' style='display: none;'><a href="NormalizationState.html#summary">summary</a></li></ul></li><li><a href="PcaState.html">PcaState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="PcaState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="PcaState.html#summary">summary</a></li></ul></li><li><a href="QualityControlState.html">QualityControlState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="QualityControlState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="QualityControlState.html#summary">summary</a></li></ul></li><li><a href="SimpleFile.html">SimpleFile</a><ul class='methods'><li data-type='method' style='display: none;'><a href="SimpleFile.html#buffer">buffer</a></li><li data-type='method' style='display: none;'><a href="SimpleFile.html#content">content</a></li><li data-type='method' style='display: none;'><a href="SimpleFile.html#name">name</a></li><li data-type='method' style='display: none;'><a href="SimpleFile.html#size">size</a></li></ul></li><li><a href="SnnGraphClusterState.html">SnnGraphClusterState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="SnnGraphClusterState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="SnnGraphClusterState.html#summary">summary</a></li></ul></li><li><a href="SummarizedExperimentDataset.html">SummarizedExperimentDataset</a><ul class='methods'><li data-type='method' style='display: none;'><a href="SummarizedExperimentDataset.html#.format">format</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperimentDataset.html#.unserialize">unserialize</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperimentDataset.html#abbreviate">abbreviate</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperimentDataset.html#annotations">annotations</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperimentDataset.html#clear">clear</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperimentDataset.html#load">load</a></li><li data-type='method' style='display: none;'><a href="SummarizedExperimentDataset.html#serialize">serialize</a></li></ul></li><li><a href="TenxHdf5Dataset.html">TenxHdf5Dataset</a><ul class='methods'><li data-type='method' style='display: none;'><a href="TenxHdf5Dataset.html#.format">format</a></li><li data-type='method' style='display: none;'><a href="TenxHdf5Dataset.html#.unserialize">unserialize</a></li><li data-type='method' style='display: none;'><a href="TenxHdf5Dataset.html#abbreviate">abbreviate</a></li><li data-type='method' style='display: none;'><a href="TenxHdf5Dataset.html#annotations">annotations</a></li><li data-type='method' style='display: none;'><a href="TenxHdf5Dataset.html#clear">clear</a></li><li data-type='method' style='display: none;'><a href="TenxHdf5Dataset.html#load">load</a></li><li data-type='method' style='display: none;'><a href="TenxHdf5Dataset.html#serialize">serialize</a></li></ul></li><li><a href="TenxMatrixMarketDataset.html">TenxMatrixMarketDataset</a><ul class='methods'><li data-type='method' style='display: none;'><a href="TenxMatrixMarketDataset.html#.format">format</a></li><li data-type='method' style='display: none;'><a href="TenxMatrixMarketDataset.html#.unserialize">unserialize</a></li><li data-type='method' style='display: none;'><a href="TenxMatrixMarketDataset.html#abbreviate">abbreviate</a></li><li data-type='method' style='display: none;'><a href="TenxMatrixMarketDataset.html#annotations">annotations</a></li><li data-type='method' style='display: none;'><a href="TenxMatrixMarketDataset.html#clear">clear</a></li><li data-type='method' style='display: none;'><a href="TenxMatrixMarketDataset.html#load">load</a></li><li data-type='method' style='display: none;'><a href="TenxMatrixMarketDataset.html#serialize">serialize</a></li></ul></li><li><a href="TsneState.html">TsneState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="TsneState.html#animate">animate</a></li><li data-type='method' style='display: none;'><a href="TsneState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="TsneState.html#summary">summary</a></li></ul></li><li><a href="UmapState.html">UmapState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="UmapState.html#animate">animate</a></li><li data-type='method' style='display: none;'><a href="UmapState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="UmapState.html#summary">summary</a></li></ul></li></ul><h3>Externals</h3><ul><li><a href="external-DataFrame.html">external:DataFrame</a></li><li><a href="external-Dataset.html">external:Dataset</a></li><li><a href="external-MultiMatrix.html">external:MultiMatrix</a></li><li><a href="external-ScranMatrix.html">external:ScranMatrix</a></li></ul><h3>Global</h3><ul><li><a href="global.html#analysisDefaults">analysisDefaults</a></li><li><a href="global.html#availableReaders">availableReaders</a></li><li><a href="global.html#callScran">callScran</a></li><li><a href="global.html#configureApproximateNeighbors">configureApproximateNeighbors</a></li><li><a href="global.html#configureBatchCorrection">configureBatchCorrection</a></li><li><a href="global.html#createAnalysis">createAnalysis</a></li><li><a href="global.html#createKanaFile">createKanaFile</a></li><li><a href="global.html#formatMarkerResults">formatMarkerResults</a></li><li><a href="global.html#freeAnalysis">freeAnalysis</a></li><li><a href="global.html#guessApproximateNeighborsConfig">guessApproximateNeighborsConfig</a></li><li><a href="global.html#guessBatchCorrectionConfig">guessBatchCorrectionConfig</a></li><li><a href="global.html#initialize">initialize</a></li><li><a href="global.html#kanaFormatVersion">kanaFormatVersion</a></li><li><a href="global.html#loadAnalysis">loadAnalysis</a></li><li><a href="global.html#parseKanaFile">parseKanaFile</a></li><li><a href="global.html#promoteToNumber">promoteToNumber</a></li><li><a href="global.html#readLines2">readLines2</a></li><li><a href="global.html#readTable2">readTable2</a></li><li><a href="global.html#retrieveParameters">retrieveParameters</a></li><li><a href="global.html#runAnalysis">runAnalysis</a></li><li><a href="global.html#saveAnalysis">saveAnalysis</a></li><li><a href="global.html#setCellLabellingDownload">setCellLabellingDownload</a></li><li><a href="global.html#setCreateLink">setCreateLink</a></li><li><a href="global.html#setResolveLink">setResolveLink</a></li><li><a href="global.html#setVisualizationAnimate">setVisualizationAnimate</a></li><li><a href="global.html#subsetInputs">subsetInputs</a></li><li><a href="global.html#summarizeArray">summarizeArray</a></li><li><a href="global.html#terminate">terminate</a></li><li><a href="global.html#validateAnnotations">validateAnnotations</a></li></ul>
</nav>

<div id="main">
    
    <h1 class="page-title">readers/mtx.js</h1>
    

    



    
    <section>
        <article>
            <pre class="prettyprint source linenums"><code>import * as scran from "scran.js";
import * as bioc from "bioconductor";
import * as afile from "./abstract/file.js";
import * as eutils from "./utils/extract.js";
import * as futils from "./utils/features.js";

/**
 * Dataset in the 10X Matrix Market format, see [here](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/matrices) for details.
 */
export class TenxMatrixMarketDataset {
    #matrix_file;
    #feature_file;
    #barcode_file;

    #dimensions;
    #raw_features;
    #raw_cells;

    /**
     * @param {SimpleFile|string|Uint8Array|File} matrixFile - A Matrix Market file.
     * On browsers, this may be a File object.
     * On Node.js, this may also be a string containing a file path.
     * @param {?(SimpleFile|string|Uint8Array|File)} featureFile - Contents of a feature annotation file.
     * If `null`, it is assumed that no file was available.
     * @param {?(SimpleFile|string|Uint8Array|File)} barcodeFile - Contents of a barcode annotation file.
     * If `null`, it is assumed that no file was available.
     */
    constructor(matrixFile, featureFile, barcodeFile) {
        if (matrixFile instanceof afile.SimpleFile) {
            this.#matrix_file = matrixFile;
        } else {
            this.#matrix_file = new afile.SimpleFile(matrixFile);
        }

        if (featureFile instanceof afile.SimpleFile || featureFile == null) {
            this.#feature_file = featureFile;
        } else {
            this.#feature_file = new afile.SimpleFile(featureFile);
        }

        if (barcodeFile instanceof afile.SimpleFile || barcodeFile == null) {
            this.#barcode_file = barcodeFile;
        } else {
            this.#barcode_file = new afile.SimpleFile(barcodeFile);
        }

        this.clear();
    }

    /**
     * Destroy caches if present, releasing the associated memory.
     * This may be called at any time but only has an effect if `cache = true` in {@linkcode TenxMatrixMarketDataset#load load} or {@linkcodeTenxMatrixMarketDataset#annotations annotations}. 
     */
    clear() {
        this.#dimensions = null;
        this.#raw_features = null;
        this.#raw_cells = null;
    }

    /**
     * @return {string} Format of this dataset class.
     * @static
     */
    static format() {
        return "MatrixMarket";
    }

    /**
     * @return {object} Object containing the abbreviated details of this dataset.
     */
    abbreviate(args) {
        var formatted = {
            "matrix": {
                name: this.#matrix_file.name(),
                size: this.#matrix_file.size()
            }
        };

        if (this.#feature_file !== null) {
            formatted.features = {
                name: this.#feature_file.name(),
                size: this.#feature_file.size()
            };
        }

        if (this.#barcode_file !== null) {
            formatted.barcodes = {
                name: this.#barcode_file.name(),
                size: this.#barcode_file.size()
            };
        }

        return formatted;
    }

    #fetch_dimensions() {
        if (this.#dimensions !== null) {
            return;
        }
        var is_gz = this.#matrix_file.name().endsWith(".gz");
        let headers = scran.extractMatrixMarketDimensions(this.#matrix_file.content(), { "compressed": is_gz });
        this.#dimensions = [headers.rows, headers.columns];
    }

    async #features() {
        if (this.#raw_features !== null) {
            return;
        }

        this.#fetch_dimensions();
        let NR = this.#dimensions[0];
        if (this.#feature_file == null) {
            this.#raw_features = new bioc.DataFrame({}, { numberOfRows: NR });
            return;
        }

        let fname = this.#feature_file.name();
        var is_gz = fname.endsWith(".gz");
        let parsed = await eutils.readTable2(this.#feature_file.content(), { compression: (is_gz ? "gz" : "none") });

        if (parsed.length == NR + 1) {
            // If it seems to have a header, we just use that directly.
            let output = {};
            let headers = parsed.shift();
            headers.forEach((x, i) => {
                output[x] = parsed.map(y => y[i]);
            });
            this.#raw_features = output;
            return;
        }

        // Otherwise, we assume it's standard 10X CellRanger output, without a header.
        if (parsed.length !== NR) {
            throw new Error("number of matrix rows is not equal to the number of rows in '" + fname + "'");
        } 

        var ids = [], symb = [];
        parsed.forEach(x => {
            ids.push(x[0]);
            symb.push(x[1]);
        });
        let output = { "id": ids, "symbol": symb };

        if (parsed[0].length >= 3) {
            let types = [];
            parsed.forEach(x => { types.push(x[2]); });
            output.type = types;
        }

        this.#raw_features = new bioc.DataFrame(output);
        return;
    }

    async #cells() {
        if (this.#raw_cells !== null) {
            return;
        }

        this.#fetch_dimensions();
        if (this.#barcode_file == null) {
            this.#raw_cells = new bioc.DataFrame({}, { numberOfRows: this.#dimensions[1] });
            return;
        }

        let bname = this.#barcode_file.name();
        var is_gz = bname.endsWith(".gz");
        let parsed = await eutils.readTable2(this.#barcode_file.content(), { compression: (is_gz ? "gz" : "none") });

        // Check if a header is present or not. Standard 10X output doesn't have a 
        // header but we'd like to support some kind of customization.
        let diff = this.#dimensions[1] - parsed.length;
        let headers;
        if (diff == 0) {
            headers = parsed[0]; // whatever, just using the first row. Hope it's unique enough!
        } else if (diff == -1) {
            headers = parsed.shift();
        } else {
            throw new Error("number of matrix columns is not equal to the number of rows in '" + bname + "'");
        }

        let annotations = {}
        headers.forEach((x, i) => {
            annotations[x] = parsed.map(y => y[i]);
        });

        for (const [k, v] of Object.entries(annotations)) {
            let conv = eutils.promoteToNumber(v);
            if (conv !== null) {
                annotations[k] = conv;
            }
        }

        this.#raw_cells = new bioc.DataFrame(annotations);
        return;
    }

    /**
     * @param {object} [options={}] - Optional parameters.
     * @param {boolean} [options.cache=false] - Whether to cache the results for re-use in subsequent calls to this method or {@linkcode TenxMatrixMarketDataset#load load}.
     * If `true`, users should consider calling {@linkcode TenxMatrixMarketDataset#clear clear} to release the memory once this dataset instance is no longer needed.
     *
     * @return {object} Object containing the per-feature and per-cell annotations.
     * This has the following properties:
     *
     * - `features`: an object where each key is a modality name and each value is a {@linkplain external:DataFrame DataFrame} of per-feature annotations for that modality.
     * - `cells`: an object containing:
     *   - `number`: the number of cells in this dataset.
     *   - `summary`: an object where each key is the name of a per-cell annotation field and its value is a summary of that annotation field,
     *     following the same structure as returned by {@linkcode summarizeArray}.
     *
     * @async
     */
    async annotations({ cache = false } = {}) {
        await this.#features();
        await this.#cells();

        let anno = {};
        for (const k of this.#raw_cells.columnNames()) {
            anno[k] = eutils.summarizeArray(this.#raw_cells.column(k));
        }

        let output = {
            "features": futils.reportFeatures(this.#raw_features, "type", cache),
            "cells": {
                "number": this.#raw_cells.numberOfRows(),
                "summary": anno
            }
        };

        if (!cache) {
            this.clear();
        }
        return output;
    }

    /**
     * @param {object} [options={}] - Optional parameters.
     * @param {boolean} [options.cache=false] - Whether to cache the results for re-use in subsequent calls to this method or {@linkcode TenxMatrixMarketDataset#annotations annotations}.
     * If `true`, users should consider calling {@linkcode TenxMatrixMarketDataset#clear clear} to release the memory once this dataset instance is no longer needed.
     *
     * @return {object} Object containing the per-feature and per-cell annotations.
     * This has the following properties:
     *
     * - `features`: an object where each key is a modality name and each value is a {@linkplain external:DataFrame DataFrame} of per-feature annotations for that modality.
     * - `cells`: a {@linkplain external:DataFrame DataFrame} containing per-cell annotations.
     * - `matrix`: a {@linkplain external:MultiMatrix MultiMatrix} containing one {@linkplain external:ScranMatrix ScranMatrix} per modality.
     * - `row_ids`: an object where each key is a modality name and each value is an integer array containing the feature identifiers for each row in that modality.
     *
     * @async
     */
    async load({ cache = false } = {}) {
        await this.#features();
        await this.#cells();

        var is_gz = this.#matrix_file.name().endsWith(".gz");
        let loaded = scran.initializeSparseMatrixFromMatrixMarket(this.#matrix_file.content(), { "compressed": is_gz });
        let output = futils.splitScranMatrixAndFeatures(loaded, this.#raw_features, "type", cache);
        output.cells = futils.cloneCached(this.#raw_cells, cache);

        if (!cache) {
            this.clear();
        }
        return output;
    }

    /**
     * @return {Array} Array of objects representing the files used in this dataset.
     * Each object corresponds to a single file and contains:
     * - `type`: a string denoting the type.
     * - `file`: a {@linkplain SimpleFile} object representing the file contents.
     */
    async serialize() {
        // These 'type's are largely retained for back-compatibility.
        let files = [ { type: "mtx", file: this.#matrix_file } ];

        if (this.#feature_file !== null) {
            files.push({ type: "genes", file: this.#feature_file });
        }

        if (this.#barcode_file !== null) {
            files.push({ type: "annotations", file: this.#barcode_file });
        }

        return files;
    }

    /**
     * @param {Array} files - Array of objects like that produced by {@linkcode TenxMatrixMarketDataset#serialize serialize}.
     * @return {TenxMatrixMarketDataset} A new instance of this class.
     * @static
     */
    static async unserialize(files) {
        let args = {};
        for (const x of files) {
            if (x.type in args) {
                throw new Error("duplicate file of type '" + x.type + "' detected during MatrixMarket unserialization");
            }
            args[x.type] = x.file;
        }

        if (!("mtx" in args)) {
            throw new Error("expected file of type 'mtx' for during MatrixMarket unserialization");
        }

        let feat = null;
        if ("genes" in args) {
            feat = args.genes;
        }

        let barcode = null;
        if ("annotations" in args) {
            barcode = args.annotations;
        }

        return new TenxMatrixMarketDataset(args.mtx, feat, barcode);
    }
}
</code></pre>
        </article>
    </section>




    
    
</div>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.6.11</a> on Fri Nov 11 2022 21:19:44 GMT+0000 (Coordinated Universal Time) using the <a href="https://github.com/clenemt/docdash">docdash</a> theme.
</footer>

<script>prettyPrint();</script>
<script src="scripts/polyfill.js"></script>
<script src="scripts/linenumber.js"></script>

<script src="scripts/search.js" defer></script>


<script src="scripts/collapse.js" defer></script>


</body>
</html>
